[Design2-119] DSCO - rework Tooltip to use React Portal functionality

levadadenys opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->
### [Design2-119] DSCO - rework Tooltip to use React Portal functionality

* Reworked `WithTooltip` to use React Portal for rendering tooltip content outside the DOM hierarchy of the parent
  component. This allows the tooltip to be rendered outside the parent component's overflow boundaries.
* `WithTooltip` is now a recomended way to use `Tooltip` and `TooltipOverlay` components. 


Here're couple screenshots that verify the positioning issue being fixed now, keep in mind code updating these tabs hasn't been added in scope of this PR and won't be since that's a separate task.
Before fix (tooltips are there, but not visible since they 'overflow' parent container):
<img width="325" alt="Знімок екрана 2024-06-06 о 16 58 44" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/57233269-1be9-4691-8eee-7ebb514c8934">

After fix:
<img width="337" alt="Знімок екрана 2024-06-06 о 16 49 48" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/7e7e8621-414b-4a8a-9de7-04ba330d1049">

## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->
- jira ticket: [Design2-119](https://codedotorg.atlassian.net/browse/DESIGN2-119)

<!--
- spec: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

bencodeorg commented:
Hey @levadadenys -- can you provide some commentary on why the accessibility issues around using React Portals noted in this thread aren't relevant here?

At a glance, I _think_ what you're doing here may be ok from an accessibility perspective, but would love to hear your general thoughts on the issue before going deeper. The thread mentions a library called "Popper" -- curious if you considered that as an alternative? Would also be great to hear if you were able to do some testing with keyboard navigation / a screen reader to see what the experience was like. Thanks!

https://github.com/facebook/react/issues/24987

sanchitmalhotra126 reviewed (commented):
TIL about React Portals! Seems really handy, but also +1 to Ben's questions around accessibility and keyboard navigation. Changes LGTM otherwise.

levadadenys commented:
Hey @bencodeorg, @sanchitmalhotra126 ,

* `Tooltips` only have text and icons as a content, meaning there'll be no tab-navigable elements inside `tooltip`
* Screen reader understands the `tooltip` value because we're passing `aria-describedBy` to the element we're attaching `tooltip` to

That being said + investigating react-bootstrap and seeing that they're using similar approach as we do now for tooltips, I would say we should be good with this approach for now. (for `Tooltips` **only (!)**).

Now, more interesting and harder question are `Popover`, `Dialog` and `Modal` DSCO components that are we going to implement in the near future (especially `Popover` since it may be used in a 'tooltip like' way and have tab-navigable elemnts).

The plan here is now that we have React Portal implementation, once I'll be working on Popovers - I check the no react-portal + Popper library option, compare these two and choose what's best for us and for the web in general. Then, if needed update `Tooltip` component.

bencodeorg commented on apps/src/componentLibrary/tooltip/Tooltip.tsx:
```diff
@@ -60,31 +62,39 @@ export const TooltipOverlay: React.FunctionComponent<TooltipOverlayProps> = ({
-    </div>
-  );
-};
+const Tooltip = React.forwardRef<HTMLDivElement, TooltipProps>(
```
Should we follow the pattern you've established previously to mark these as for internal use only and name it `_Tooltip`? I see you said in the docs that this could technically be used directly, but maybe we should just not advise people to use it directly?

levadadenys replied:
Yeah, was think about this also. Thought it might be better to keep it 'public' just because other internal components are internal cause they are not designed for any usage outside of Design System, while `Tooltip` can theoretically be used if user wants to create some custom positioned tooltip. At the same time, recommended way to use it is, of course, `WithTooltip`.


Though, now that you're asking about it, I'm now hesitant about leaving `Tooltip`  public

bencodeorg replied:
I'd vote to make it not public (ie, drive usage towards `WithTooltip` / away from custom implementations if possible), but up to you :)

bencodeorg commented on apps/src/componentLibrary/tooltip/Tooltip.story.tsx:
```diff
@@ -28,23 +28,17 @@ const SingleTemplate: StoryFn<TooltipProps> = args => (
-            style={{margin: 0}}
-            onClick={() => {}}
-          >
+        <WithTooltip key={componentArg.tooltipId} tooltipProps={componentArg}>
```
Love having these stories as documentation of how to use these components 🙌

bencodeorg commented on apps/src/componentLibrary/tooltip/Tooltip.tsx:
```diff
@@ -60,31 +62,39 @@ export const TooltipOverlay: React.FunctionComponent<TooltipOverlayProps> = ({
+  ) => {
+    return (
+      <div
+        ref={ref}
```
Unfamiliar with `forwardRef`...what does it mean to assign the ref to an HTML element attribute `ref`?

levadadenys replied:
Oh, `ref`s are awesome, though specific and not always needed feature of React! `ref` is used to access DOM element inside React code/component. Here it's used to get `Tooltip` DOM element height and width since we can't predict them in any other way.

`forwardRef` is a method of React that allows you to pass `ref` to the component as a prop. (Without forwardRef you'll be only able to set a ref to an html element inside your component, meaning you won't be able to access this dom element at current component's parent component and you won't be able to pass ref as a prop to this component)

You can read more about them here:
https://react.dev/learn/manipulating-the-dom-with-refs
https://react.dev/reference/react/forwardRef

bencodeorg replied:
Oh cool I didn't realize you could assign a ref directly to a native HTML element, although that seems now fairly common now that I look at docs/other examples in our code. Thanks!

bencodeorg commented on apps/src/componentLibrary/tooltip/WithTooltip.tsx:
```diff
@@ -13,16 +30,180 @@ const WithTooltip: React.FunctionComponent<WithTooltipProps> = ({
+        case 'onRight':
+          styles = {
+            top: `${
+              rect.top +
```
This math looks repeated / complicated with `onLeft` (maybe other repetition too?), could we extract it?

levadadenys replied:
it's `top` for `onRight` and `onLeft`,
`left` for `onTop` and `onBottom`,
`left` `ltr` for `onRight` and `left` `rtl` for `onLeft`
`left` `rtl` for `onRight` and `left` `ltr` for `onLeft`.

I agree this switch case is now bulky, though not sure about the best way of refactoring it yet. Will think more and try updating

bencodeorg replied:
Lines 74-77 and 105-109 are exactly the same, yeah? I think just at least pulling out exactly copied math would be good, not sure if you need to refactor to fancily do all math with single helper or anything.

bencodeorg reviewed (approved):
Thanks for the notes about accessibility! Agreed that I think we're covered in this case where there aren't potential tab navigation issues, and looks like you've done a nice job with the aria attributes here!

levadadenys commented:
Updated the PR, cleaned up the tooltip positioning calculation logic to be more readable and cleaner.
PR is ready to be merged now.

Tooltips will be positioned correctly now, no matter of the size of the container of element triggering tooltip show. To do that it's required that `componentToWrap` (component we're going to attach tooltip to) will support `onFocus, onBlur, onMouseEnter, onMouseLeave` props. It's a known issue that know `WIthTooltip` won't work with components that doesn't pass these props to actual html element (e.g. Design System components right now). Planning to work on it in scope of a follow up PR. The idea here is either add support of native html element props to existing Design System components since we might actually need them anyways (?) at some point OR use `Popper` lib if that'll be easier, though the question about adding `onFocus, onBlur, onMouseEnter, onMouseLeave` and all other general html attributes/props to all the Design System components still remains active

cc @bencodeorg @sanchitmalhotra126

bencodeorg commented:
Much easier to understand to read the positioning math now, thanks for extracting!

Not blocking, but re: checking out what gets passed as a child to `WithTooltip`, could we do something like this (although I think throwing an error is a little harsh, maybe just a warning):

https://github.com/code-dot-org/code-dot-org/blob/38f42be85693508bdaa6c3eabf2421fe44e2e37b/apps/src/javalab/components/JavalabDropdown.jsx#L14-L26

levadadenys commented:
> Much easier to understand to read the positioning math now, thanks for extracting!
> 
> Not blocking, but re: checking out what gets passed as a child to `WithTooltip`, could we do something like this (although I think throwing an error is a little harsh, maybe just a warning):

@bencodeorg 
Not sure I understand. Do you mean add a check that children is `button` specifically? But the idea here is that almost everything can be wrapped up with `withTooltip`, meaning you should be able to attach `tooltip` to almost everything. Are storybook stories now are now making people assume it can be used only with buttons? I might need to do some additional work on stories and documentation then.

Merging PR now as is, will make additional changes in a follow up PR if needed.
